File: examples/typescript/client_goal_templates.ts
// ekoDB TypeScript Client - Goal Template CRUD Example
//
// Demonstrates creating, listing, getting, updating, and deleting goal templates.

import { EkoDBClient } from "@ekodb/ekodb-client";

async function main() {
  console.log("=== ekoDB Goal Template CRUD Example (TypeScript) ===\n");

  const baseUrl = process.env.API_BASE_URL || "http://localhost:8080";
  const apiKey = process.env.API_BASE_KEY;
  if (!apiKey) throw new Error("API_BASE_KEY environment variable is required");

  const client = new EkoDBClient(baseUrl, apiKey);

  // 1. Create a goal template
  console.log("--- Creating goal template ---");
  const template = await client.goalTemplateCreate({
    title: "Data Migration",
    description: "Template for migrating data between schemas",
    steps: [
      { description: "Analyze source schema" },
      { description: "Create target schema" },
      { description: "Migrate records" },
      { description: "Validate results" },
    ],
  });
  const templateId = template.id;
  console.log(`Created template: ${template.title} (id: ${templateId})`);

  // 2. List all templates
  console.log("\n--- Listing templates ---");
  const list = await client.goalTemplateList();
  console.log("Templates:", list);

  // 3. Get template by ID
  console.log("\n--- Getting template ---");
  const fetched = await client.goalTemplateGet(templateId);
  console.log(`Fetched: ${fetched.title}`);

  // 4. Update template
  console.log("\n--- Updating template ---");
  const updated = await client.goalTemplateUpdate(templateId, {
    description: "Updated: comprehensive data migration workflow",
  });
  console.log(`Updated description: ${updated.description}`);

  // 5. Delete template
  console.log("\n--- Deleting template ---");
  await client.goalTemplateDelete(templateId);
  console.log("Template deleted successfully");

  console.log("\n✓ Goal template CRUD example completed");
}

main().catch(console.error);

File: examples/python/client_goal_templates.py
"""ekoDB Python Client - Goal Template CRUD Example

Demonstrates creating, listing, getting, updating, and deleting goal templates.
"""

import asyncio
import os

from ekodb_client import Client


async def main():
    print("=== ekoDB Goal Template CRUD Example (Python) ===\n")

    base_url = os.getenv("API_BASE_URL", "http://localhost:8080")
    api_key = os.getenv("API_BASE_KEY")
    if not api_key:
        raise ValueError("API_BASE_KEY environment variable is required")

    client = Client.new(base_url, api_key)

    # 1. Create a goal template
    print("--- Creating goal template ---")
    template = await client.goal_template_create(
        {
            "title": "Data Migration",
            "description": "Template for migrating data between schemas",
            "steps": [
                {"description": "Analyze source schema"},
                {"description": "Create target schema"},
                {"description": "Migrate records"},
                {"description": "Validate results"},
            ],
        }
    )
    template_id = template["id"]
    print(f"Created template: {template.get('title')} (id: {template_id})")

    # 2. List all templates
    print("\n--- Listing templates ---")
    templates = await client.goal_template_list()
    print(f"Templates: {templates}")

    # 3. Get template by ID
    print("\n--- Getting template ---")
    fetched = await client.goal_template_get(template_id)
    print(f"Fetched: {fetched.get('title')}")

    # 4. Update template
    print("\n--- Updating template ---")
    updated = await client.goal_template_update(
        template_id,
        {
            "description": "Updated: comprehensive data migration workflow",
        },
    )
    print(f"Updated description: {updated.get('description')}")

    # 5. Delete template
    print("\n--- Deleting template ---")
    await client.goal_template_delete(template_id)
    print("Template deleted successfully")

    print("\n✓ Goal template CRUD example completed")


if __name__ == "__main__":
    asyncio.run(main())
